### PR TITLE
core: try to solve warnings caused by Apple's new LAPACK interface

### PR DESCRIPTION
**Pull request by fengyuentau**
Resolves https://github.com/opencv/opencv/issues/24660

Apple's BLAS documentation: https://developer.apple.com/documentation/accelerate/blas?language=objc

New interface since macOS >= 13.3, iOS >= 16.4.

Todo:

- [x] Detect macOS version.
- [x] ~Detect iOS versions (major and minor version).~ No calling of Accelerate New LAPACK on iOS.
- [x] Solve calling `cblas_cgemm` and `cblas_zgemm`.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
See https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_DEPLOYMENT_TARGET.html for the first item. Looks like it's what you need.

**Comment by asmorkalov** [comment]
What is the issue with the second item?

**Comment by asmorkalov** [comment]
@fengyuentau friendly reminder.

**Comment by fengyuentau** [comment]
> @fengyuentau friendly reminder.

Let me get back on this item later.

**Comment by fengyuentau** [comment]
I am trying to make this PR work but currently stuck with "Core_GEMM.accurarcy". I found that the step of complex mat does not reflect the inner dimension of the mat, but yet calling blas routines give correct result. See cases with `CV_32FC2/CV_64FC2` type below for example.

```
[ RUN      ] Core_GEMM.accuracy
A.size=282 x 221, A.channels()=1, A.step=884, A.step1=221
B.size=221 x 74, B.channels()=1, B.step=296, B.step1=74
dst.size=282 x 74, dst.channels()=1, dst.step=296, dst.step1=74
Type is CV_32FC1
src1_step=884, ldsrc1=221, src2_step=296, ldsrc2=74, dst_step=296, lddst=74, sizeof(float)=4

A.size=638 x 3, A.channels()=2, A.step=24, A.step1=6
B.size=18 x 638, B.channels()=2, B.step=5128, B.step1=1282
dst.size=3 x 18, dst.channels()=2, dst.step=200, dst.step1=50
Type is CV_32FC2
src1_step=24, ldsrc1=3, src2_step=5128, ldsrc2=641, dst_step=200, lddst=25, sizeof(std::complext<float>)=8

A.size=1 x 36, A.channels()=1, A.step=288, A.step1=36
B.size=6 x 1, B.channels()=1, B.step=8, B.step1=1
dst.size=36 x 6, dst.channels()=1, dst.step=48, dst.step1=6
Type is CV_64FC1

A.size=3 x 467, A.channels()=2, A.step=7488, A.step1=936
B.size=2 x 467, B.channels()=2, B.step=7472, B.step1=934
dst.size=3 x 2, dst.channels()=2, dst.step=32, dst.step1=4
Type is CV_64FC2

A.size=7 x 6, A.channels()=1, A.step=36, A.step1=9
B.size=735 x 6, B.channels()=1, B.step=24, B.step1=6
dst.size=7 x 735, dst.channels()=1, dst.step=2972, dst.step1=743
Type is CV_32FC1

A.size=16 x 76, A.channels()=1, A.step=308, A.step1=77
B.size=16 x 184, B.channels()=1, B.step=772, B.step1=193
dst.size=76 x 184, dst.channels()=1, dst.step=752, dst.step1=188
Type is CV_32FC1

A.size=37 x 989, A.channels()=2, A.step=7912, A.step1=1978
B.size=37 x 220, B.channels()=2, B.step=1800, B.step1=450
dst.size=989 x 220, dst.channels()=2, dst.step=1784, dst.step1=446
Type is CV_32FC2

A.size=84 x 13, A.channels()=1, A.step=104, A.step1=13
B.size=13 x 264, B.channels()=1, B.step=2176, B.step1=272
dst.size=84 x 264, dst.channels()=1, dst.step=2112, dst.step1=264
Type is CV_64FC1

A.size=190 x 162, A.channels()=2, A.step=2592, A.step1=324
B.size=162 x 316, B.channels()=2, B.step=5056, B.step1=632
dst.size=190 x 316, dst.channels()=2, dst.step=5072, dst.step1=634
Type is CV_64FC2
src1_step=2592, ldsrc1=162, src2_step=5056, ldsrc2=316, dst_step=5072, lddst=317, sizeof(std::complext<double>)=16

A.size=430 x 2, A.channels()=1, A.step=80, A.step1=10
B.size=2 x 379, B.channels()=1, B.step=3104, B.step1=388
dst.size=430 x 379, dst.channels()=1, dst.step=3032, dst.step1=379
Type is CV_64FC1
src1_step=80, ldsrc1=10, src2_step=3104, ldsrc2=388, dst_step=3032, lddst=379, sizeof(double)=8

A.size=974 x 26, A.channels()=2, A.step=208, A.step1=52
B.size=1 x 974, B.channels()=2, B.step=7808, B.step1=1952
dst.size=26 x 1, dst.channels()=2, dst.step=64, dst.step1=16
Type is CV_32FC2
src1_step=208, ldsrc1=26, src2_step=7808, ldsrc2=976, dst_step=64, lddst=8, sizeof(std::complext<float>)=8

A.size=2 x 343, A.channels()=1, A.step=1376, A.step1=344
B.size=412 x 2, B.channels()=1, B.step=40, B.step1=10
dst.size=343 x 412, dst.channels()=1, dst.step=1648, dst.step1=412
Type is CV_32FC1

A.size=5 x 4, A.channels()=2, A.step=64, A.step1=8
B.size=141 x 4, B.channels()=2, B.step=80, B.step1=10
dst.size=5 x 141, dst.channels()=2, dst.step=2336, dst.step1=292
Type is CV_64FC2

A.size=11 x 56, A.channels()=1, A.step=448, A.step1=56
B.size=48 x 56, B.channels()=1, B.step=448, B.step1=56
dst.size=11 x 48, dst.channels()=1, dst.step=408, dst.step1=51
Type is CV_64FC1

A.size=25 x 725, A.channels()=2, A.step=11600, A.step1=1450
B.size=25 x 936, B.channels()=2, B.step=14976, B.step1=1872
dst.size=725 x 936, dst.channels()=2, dst.step=14976, dst.step1=1872
Type is CV_64FC2
/Users/fengyuantao/Workspace/fytao/workspace/fix_lapack_warnings/opencv/modules/ts/src/ts.cpp:618: Failure
Failed

        failure reason: Arithmetic exception
        test case #14
        seed: 00000000000c5a6a
-----------------------------------
        LOG:
General failure:
        Arithmetic exception (-6)

-----------------------------------
        CONSOLE: ........
-----------------------------------

[  FAILED  ] Core_GEMM.accuracy (91 ms)
[----------] 1 test from Core_GEMM (91 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (91 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Core_GEMM.accuracy

 1 FAILED TEST
```

Could you guide me the right way to understand this? @vpisarev @asmorkalov @opencv-alalek

File: cmake/OpenCVFindLAPACK.cmake [hunk 1]
```diff
@@ -106,11 +106,37 @@ macro(ocv_lapack_check)
       list(APPEND __link_directories ${LAPACK_LINK_LIBRARIES})
     endif()
 
+    set(LAPACK_TRY_COMPILE_DEF "")
+    if(LAPACK_IMPL STREQUAL "LAPACK/Apple") # https://github.com/opencv/opencv/issues/24660
+      # TODO: Need to find a way detecting macOS / iOS versions
+      #       enbale if macOS >= 13.3 or iOS >= 16.4
```
Comment (asmorkalov) [inline cmake/OpenCVFindLAPACK.cmake:112]: to-do is not relevant any more.

File: cmake/OpenCVFindLAPACK.cmake [hunk 2]
```diff
@@ -106,11 +106,37 @@ macro(ocv_lapack_check)
...
+
+      # Extract major and minor versions
+      string(REGEX MATCHALL "[0-9]+" MACOS_VERSION_NUMBER_LIST ${MACOS_VERSION})
+      list(LENGTH MACOS_VERSION_NUMBER_LIST MACOS_VERSION_NUMBER_LIST_LENGTH)
+      if(MACOS_VERSION_NUMBER_LIST_LENGTH GREATER_EQUAL 2)
+        list(GET MACOS_VERSION_NUMBER_LIST 0 MACOS_MAJOR_VERSION)
+        list(GET MACOS_VERSION_NUMBER_LIST 1 MACOS_MINOR_VERSION)
+      endif()
+      # Enable Accelerate New LAPACK if macOS >= 13.3
+      if(MACOS_MAJOR_VERSION GREATER_EQUAL 13 AND MACOS_MINOR_VERSION GREATER_EQUAL 3)
```
Comment (asmorkalov) [inline cmake/OpenCVFindLAPACK.cmake:127]: It does not work for 14.2, for example. Please use special version comparison: https://cmake.org/cmake/help/latest/command/if.html#version-comparisons. Example: https://github.com/opencv/opencv/blob/f9dd20eb07210f4759889f667fc8fc1ad90f55da/CMakeLists.txt#L865

Comment (fengyuentau) [reply cmake/OpenCVFindLAPACK.cmake]: Done.

File: cmake/OpenCVFindLAPACK.cmake (hunk 1 above)
Comment (fengyuentau) [reply cmake/OpenCVFindLAPACK.cmake]: Done. I do not want to mess around with ios build script, so skipping this for ios.

**Comment by fengyuentau** [comment]
@vpisarev @asmorkalov @opencv-alalek Could you review and merge if no further changes are needed?

File: modules/core/src/hal_internal.cpp [hunk 3]
```diff
@@ -111,8 +111,17 @@ set_value(fptype *dst, size_t dst_ld, fptype value, size_t m, size_t n)
...
+    long* piv = new long[m];
+    long lda = (long)(a_step / sizeof(fptype));
+    long _m = static_cast<long>(m), _n = static_cast<long>(n);
+    long _info[1];
+#else
     int* piv = new int[m];
+    int lda = (int)(a_step / sizeof(fptype));
+    int _m = m, _n = n;
+    int* _info = info;
+#endif
```
[the remark this answers is not in the document]
Comment (fengyuentau) [reply modules/core/src/hal_internal.cpp]: Done.

File: modules/core/src/hal_internal.cpp [hunk 4]
```diff
@@ -210,11 +231,23 @@ lapack_Cholesky(fptype* a, size_t a_step, int m, fptype* b, size_t b_step, int n
...
+    long* iworkBuf = new long[8*std::min(m, n)];
+#else
+    int _m = m, _n = n;
+    int* _info = info;
     int lda = (int)(a_step / sizeof(fptype));
     int ldv = (int)(v_step / sizeof(fptype));
     int ldu = (int)(u_step / sizeof(fptype));
     int lwork = -1;
     int* iworkBuf = new int[8*std::min(m, n)];
+#endif
```
[the remark this answers is not in the document]
Comment (fengyuentau) [reply modules/core/src/hal_internal.cpp]: Done.

**Review by asmorkalov — APPROVED** [review]

File: cmake/OpenCVFindLAPACK.cmake [hunk 5]
```diff
@@ -108,11 +108,27 @@ macro(ocv_lapack_check)
...
-      set(LAPACK_TRY_COMPILE_DEF "-DACCELERATE_NEW_LAPACK")
       # TODO: Need to find a way detecting macOS / iOS versions
       #       enbale if macOS >= 13.3 or iOS >= 16.4
-      # add_compile_definitions(ACCELERATE_NEW_LAPACK)
-      # add_compile_definitions(ACCELERATE_LAPACK_ILP64)
+
+      # Get macOS version
+      execute_process(COMMAND sw_vers -productVersion
+                      OUTPUT_VARIABLE MACOS_VERSION
+                      OUTPUT_STRIP_TRAILING_WHITESPACE)
```
Comment (seanm) [inline cmake/OpenCVFindLAPACK.cmake:114]: This is a problem... it checks the version of macOS on the machine doing the compiling, it does not check the version of macOS that's actually running the code (at runtime). It's common for a developer to be using a new macOS but wanting his code to still run on an old macOS, and that's why these kinds of checks are inappropriate.

Instead, we must check against the '[deployment target](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_DEPLOYMENT_TARGET.html)' chosen by the developer, set in `CMAKE_OSX_DEPLOYMENT_TARGET`.

Could we revert this merge, or correct this quickly?

Comment (fengyuentau) [reply cmake/OpenCVFindLAPACK.cmake:114]: Let me take care of this. cc @asmorkalov @opencv-alalek

**Pull request merged**